### PR TITLE
do not allow client password logins if token auth is enforced or 2FA …

### LINKED ISSUE #24768: Forbid password login on APIs if user is two-factor authenticated
### Steps to reproduce
1. Install 2FA app
2. Access DAV/OCS whatever with username+password

### Expected behaviour
Access should be forbidden unless the password is a token.

### Actual behaviour
Login is successful.

ref https://github.com/owncloud/core/pull/24559

### LINKED ISSUE #24779: For all new installed ownClouds enforce application specific passwords
see #23458 
### Steps to reproduce
1. Install ownCloud
2. Create user
3. Set up (sync) client with username + password
### Expected behaviour

:boom: Auth should fail.
### Actual behaviour

Auth by password is still allowed.

cc @LukasReschke @PVince81

### PR DESCRIPTION
**Pull request by ChristophWurst**
…is enabled

fixes #24779
fixes #24768

TODO
- [x] Check if instance enforces token auth (default for new installations)
- [x] Check if 2FA is enforced for the user
- [x] Add config to config.sample.php

Tested with DAV and OCS, works as expected.

### CODE REVIEW

File: lib/private/User/Session.php [hunk 1]
```diff
@@ -89,12 +93,15 @@ class Session implements IUserSession, Emitter {
 	 * @param ISession $session
 	 * @param ITimeFactory $timeFacory
 	 * @param IProvider $tokenProvider
+	 * @param IConfig $config
+	 * @param TwoFactorAuthManager
 	 */
-	public function __construct(IUserManager $manager, ISession $session, ITimeFactory $timeFacory, $tokenProvider) {
+	public function __construct(IUserManager $manager, ISession $session, ITimeFactory $timeFacory, $tokenProvider, IConfig $config) {
```
Comment (nickvergessen) [inline lib/private/User/Session.php]: $tokenProvider has no type hint, intended?

File: lib/private/User/Session.php [hunk 2]
```diff
@@ -89,12 +93,15 @@ class Session implements IUserSession, Emitter {
 	 * @param ISession $session
 	 * @param ITimeFactory $timeFacory
 	 * @param IProvider $tokenProvider
+	 * @param IConfig $config
+	 * @param TwoFactorAuthManager
```
Comment (nickvergessen) [inline lib/private/User/Session.php]: nop

File: lib/private/User/Session.php [hunk 3]
```diff
@@ -327,6 +334,60 @@ public function login($uid, $password) {
...
 
+	/**
+	 * Tries to log in a client
+	 *
+	 * Checks token auth enforced
+	 * Checks 2FA enabled
+	 *
+	 * @param string $user
+	 * @param string $password
+	 * @throws LoginException
```
Comment (nickvergessen) [inline lib/private/User/Session.php]: Are you sure?

File: lib/private/User/Session.php (hunk 1 above)
Comment (ChristophWurst) [inline lib/private/User/Session.php]: yes, because ``null`` in injected during setup

File: lib/private/User/Session.php (hunk 3 above)
Comment (ChristophWurst) [inline lib/private/User/Session.php]: yes, login() throws exceptions

Comment (nickvergessen) [inline lib/private/User/Session.php]: ah missed the return line

**Comment by ChristophWurst** [comment]
@PVince81 @nickvergessen @schiesbn @rullzer @LukasReschke test & review please :-)

Tested also with LDAP, seems to work :speak_no_evil:

File: lib/private/Setup.php [hunk 4]
```diff
@@ -362,6 +362,9 @@ public function install($options) {
 			$config->setAppValue('core', 'installedat', microtime(true));
 			$config->setAppValue('core', 'lastupdatedat', microtime(true));
 
+			// Disable strict token auth by default
+			$config->setSystemValue('token_auth_enforced', false);
```
Comment (PVince81) [inline lib/private/Setup.php]: I'd remove that, let it rely on the default value being false anyway

File: lib/private/Updater.php [hunk 5]
```diff
@@ -265,6 +265,9 @@ private function doUpgrade($currentVersion, $installedVersion) {
 				$this->emit('\OC\Updater', 'failure', [$appId . ': ' . $exception->getMessage()]);
 			}
 
+			// Disable strict token auth by default
+			$this->config->setSystemValue('token_auth_enforced', false);
```
Comment (PVince81) [inline lib/private/Updater.php]: same here, remove it.

**Comment by PVince81** [comment]
Looks good :+1:

**Comment by PVince81** [comment]
Second review ? @rullzer @nickvergessen

**Comment by rullzer** [comment]
Awesome!
Tested and works over here.
Code looks good.

:+1:

**Comment by ChristophWurst** [comment]
Failing test looks unrelated
```
16:27:55 There was 1 failure:
16:27:55 
16:27:55 1) OCA\Files_Sharing\Controllers\ShareControllerTest::testShowShare
16:27:55 Failed asserting that two objects are equal.
16:27:55 --- Expected
16:27:55 +++ Actual
16:27:55 @@ @@
16:27:55          'previewEnabled' => true
16:27:55 +        'previewMaxX' => null
16:27:55 +        'previewMaxY' => null
16:27:55      )
16:27:55      'renderAs' => 'base'
16:27:55      'appName' => 'files_sharing'
16:27:55      'headers' => Array (...)
16:27:55      'cookies' => Array ()
16:27:55      'status' => 200
16:27:55      'lastModified' => null
16:27:55      'ETag' => null
16:27:55      'contentSecurityPolicy' => OCP\AppFramework\Http\ContentSecurityPolicy Object (...)
16:27:55  )
```

**Comment by PVince81** [comment]
@ChristophWurst likely due to this PR: https://github.com/owncloud/core/pull/24346

Will rerun the tests on master and make tickets so we can fix these today.

**Pull request merged**